OS/Atmega32/src/SOS.c

```diff
@@ -0,0 +1,441 @@
+   
+   /* Install ISR */
+   Interrupt_Install(TIMER2_OVF_VECTOR_NUMBER, Systick_ISR);
+   
+   /* Enable Global interrupts. */
+   INTERRUPTS_Enable();
+}
+
+/**
+* @brief: This function is the schedular of the OS.
+*/
+void SOS_Run(void)
+{
+   /* Start the OS clock. */
+   TCCR2 |= gstr_SYSTICK_Config.u8_prescale;
+   
+   /* Updates Task states before running. */
+   SOS_UpdateStates();
+   
+   while(1)
+   {
+      /* Variables to store the next running task. */
+      uint8_t u8_TaskID = IDLE_TASK;
+      uint8_t u8_TaskPriority = MAX_TASK_PRIORTITY;
+      
+      /* Loop on all tasks */
+      for(uint8_t u8_TaskCounter = 0; u8_TaskCounter< gu8_CreatedTasks; u8_TaskCounter++)
+      {
+         /* If the task is ready and lower priority. */
+         if((SOS_TASK_READY == gastr_Tasks[u8_TaskCounter].u8_State) && (u8_TaskPriority > gastr_Tasks[u8_TaskCounter].u8_Priority) )
+         {
+            /* Set the task id and it's priority. */
+            u8_TaskID = u8_TaskCounter;
+            u8_TaskPriority = gastr_Tasks[u8_TaskCounter].u8_Priority;
+         }
+      }
+      
+      /* If a Task is selected. */
+      if(IDLE_TASK != u8_TaskID)
+      {
+         /* Change its state to running. */
+         gastr_Tasks[u8_TaskID].u8_State = SOS_TASK_RUNNING;
+         #if CALCULATE_CPU_LOAD
+            /* Set CPU state to working. */
+            CPU_State = CPU_WORKING;
+         #endif
+         /* Run Task. */
+         gastr_Tasks[u8_TaskID].ptr_Task();
+         /* Set State waiting. */
+         gastr_Tasks[u8_TaskID].u8_State = SOS_TASK_WAITING;
+      }
+      /* If No task is selected. */
+      else
+      {
+         /* Run Idle Task */
+         Idle_Task();
+      }
+   }
+}
+
+/**
+* @brief: This function Creates a Task.
+*
+* @param [in]  Task        -  Pointer to the task function.
+* @param [in]  Priority    -  Priority of the Task.
+* @param [in]  Periodicity -  Periodicity of the task.
+* @param [out] Handler     -  Handler of the task.
+*
+* @return function error state.
+*/
+ERROR_STATE_t SOS_CreateTask(Ptr_VoidFuncVoid_t Task, uint8_t Priority, uint16_t Periodicity, STR_TaskHandler_t * Handler)
+{
+   /* variable to store function error state. */
+   ERROR_STATE_t Create_ErrorState;
+   
+   /* Check if Max tasks are created. */
+   if(gu8_CreatedTasks == (MAX_TASKS_NUMBER))
+   {
+      /* Set Function Error State To Full Task. */
+      Create_ErrorState = (SOS_MODULE_ID | E_TASKS_FULL);
+   }
+   /* Validate valid pointer to task. */
+   else if(NULL_PTR == Task)
+   {
+      /* Set Function Error State To Null Pointer. */
+      Create_ErrorState = (SOS_MODULE_ID | E_NULL_PTR);
+   }
+   /* Validate Valid priority. */
+   else if(MAX_TASK_PRIORTITY < Priority)
+   {
+      /* Set Function Error State To Invalid Priority. */
+      Create_ErrorState = (SOS_MODULE_ID | E_INVALID_PRIORITY);
+   }
+   else
+   {
+      /* Initialize Task data. */
+      gastr_Tasks[gu8_CreatedTasks].u8_State = SOS_TASK_WAITING;
```

Comment:
I think, you should make initial state of the task to ready, because by default all tasks should be ready at T=0, but if you make them waiting means that from T=0 till T=first task period, the system is idle.